### PR TITLE
examples: dual_qemu_ivhshmem: add RPMsg over IVSHMEM sample code.

### PR DESCRIPTION
**Pull request by uLipe**
The additions include a backend to glue the Zephyr IVSHMEM device driver into the openAMP code making it usable to send data between two QEMU instances using the RPMsg protocol.

Also a custom shell command in the host side application is provided to send string messages for a number of times allowing to perform stress test of the communication.

This sample has a limitation regarding the order of initialization of the QEMU host and side instances, the host side must be initialized first, followed by the remote side please refer the README of this sample for more info.

### CODE REVIEW

**Comment by uLipe** [comment]
This PR is a work in progress but I would like to share with even in this state in order to give a follow up of this work.

File: west.yml [hunk 1]
```diff
@@ -5,14 +5,19 @@ manifest:
   remotes:
     - name: openamp
       url-base: https://github.com/OpenAMP
+    # Temporary solution until openamp-zephyr-staging gets the 3.4 additions
+    # They are needed to operate samples based on the IVSHMEM.
+    - name: zephyr-upstream
+      url-base: https://github.com/zephyrproject-rtos
```
Comment (uLipe) [inline west.yml:11]: At this point the additions from the 3.4 branch were not available into the zephyr staging project from the openAMP repository, so this is a temporary solution that will be removed once 3.4 becomes available.

The 3.4 branch is needed in order to have IVSHMEM driver and some fixes for the PCIe controller for ARM.

File: examples/zephyr/rpmsg_over_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+	shmem_base += VDEV_STATUS_SIZE;
+
+	LOG_DBG("Memory got from ivshmem: %p, size: %ld \n", (void *)shmem_base, shmem_size);
+
+	shm_device.regions[0].virt = (void *)(shmem_base);
+	shm_device.regions[0].size = shmem_size;
+	shm_physmap[0] = (metal_phys_addr_t)(shmem_base);
+
+
+#ifdef CONFIG_OPENAMP_MASTER
```
Comment (uLipe) [inline examples/zephyr/rpmsg_over_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c:211]: This is something that needs to be changed in zephyr itself, I know this symbol is deprecated in the openAMP code base, but Zephyr module interface KCONFIG redefines this symbol and it is used in any glue code between Zephyr and openAMP.

**Comment by uLipe** [comment]
@wmamills , @danmilea , @arnopo just tagging you to follow-up, but feel free to review and reproduce this one.

File: examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+static void virtio_set_features(struct virtio_device *vdev,
+				uint32_t features)
+{
+}
+
+static void virtio_notify(struct virtqueue *vq)
+{
+	uint16_t peer_dest_id = ivshmem_get_id(ivshmem_dev)
+#ifdef CONFIG_OPENAMP_MASTER
+	+1;
```
Comment (arnopo) [inline examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c:118]: nitpicking  
`peer_dest_id += 1; `would be more readable
Else to add a space between '+' an '1'

File: west.yml [hunk 4]
```diff
@@ -14,5 +14,5 @@ manifest:
   projects:
     - name: openamp-zephyr-staging
       path: zephyr
-      revision: v3.2-branch
-      import: true
+      revision: v3.4-branch
+      import: true
```
Comment (arnopo) [inline west.yml:18]: Would be better to update west.yml in a separate patch.

**Review by arnopo — COMMENTED** [review]
2 minor comments else LGTM

**Review by arnopo — APPROVED** [review]
It seems that a coding rule in Zephyr is to put the code in the `src` folder.  This could be the case for the rpmsg_ivshmem_backend folder, but keeping it that way works for me too.

File: examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c [hunk 5]
```diff
@@ -0,0 +1,314 @@
+/*
+ * Copyright (c) 2023, Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define LOG_MODULE_NAME	rpmsg_ivshmem_backend
+#define LOG_LEVEL CONFIG_IVSHMEM_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
```
Comment (TanmayShah-xilinx) [inline examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c:10]: can we move LOG_* macros after header files ? This way we don't have to include log.h file separately.

File: examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c [hunk 6]
```diff
@@ -0,0 +1,314 @@
+/*
+ * Copyright (c) 2023, Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
```
Comment (TanmayShah-xilinx) [inline examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c:4]: I don't know too much about including license information, but above main.c has BSD-3-Clause, and here Apache-2.0. I don't know the difference, but should we unify licenses for all source files? @arnopo, @wmamills what is the guideline here?

File: examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+
+#include <zephyr/kernel.h>
+#include <zephyr/init.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/virtualization/ivshmem.h>
+#include <zephyr/sys/printk.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
```
Comment (TanmayShah-xilinx) [inline examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c:20]: standard header files at the top. @arnopo Do we follow coding style from openamp library or zephyr RTOS coding style?

**Review by TanmayShah-xilinx — COMMENTED** [review]
Hello,

Thanks for your patch.
minor comments related to coding style, otherwise looks good to me.

Thanks,
Tanmay

File: examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c (hunk 6 above)
Comment (arnopo) [reply examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c:4]: Zephyr is apache-2.0, 
openamp is BSD-3

I would say as it is Zephyr application, would be better to be Apache 2.0

File: examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c (hunk 5 above)
Comment (uLipe) [reply examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c:10]: This pattern is the required pattern from Zephyr for registering a LOG capable module, the both defines are consumed by the header file and then in the order, by the LOG registering macro.

File: examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c (hunk 6 above)
Comment (arnopo) [reply examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c:4]: @uLipe : Please could you replace BSD by Apache license in your dev?

Comment (uLipe) [reply examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c:4]: Sure, thanks for clarify.

File: examples/zephyr/dual_qemu_ivshmem/README.rst [hunk 8]
```diff
@@ -0,0 +1,241 @@
+OpenAMP RPMSG over IVSHMEM system reference sample
```
[the remark this answers is not in the document]
Comment (arnopo) [reply examples/zephyr/dual_qemu_ivshmem/README.rst:1]: based on  other in-tree Zephyr examples, it is not necessary to add a License here

File: examples/zephyr/dual_qemu_ivshmem/host/boards/pcie_ivshmem.dtsi [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#include <zephyr/dt-bindings/pcie/pcie.h>
+
+/ {
+	ivhsmem {
+		ivshmem0: ivshmem {
+			compatible = "qemu,ivshmem";
+
+			vendor-id = <0x1af4>;
```
[the remark this answers is not in the document]
Comment (uLipe) [reply examples/zephyr/dual_qemu_ivshmem/host/boards/pcie_ivshmem.dtsi:13]: I searched for this vendor ID in the PCI-SIG group list and it did not return results, it seems this was just assigned randomly on the zephyr in-tree sample:

https://github.com/zephyrproject-rtos/zephyr/blob/690b21718ae67a972b442dc98c9eccc3f140df0c/samples/drivers/virtualization/ivshmem/doorbell/boards/pcie_ivshmem.dtsi#L13

File: examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c (hunk 6 above)
Comment (arnopo) [reply examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c:4]: It seems that there are still other headers with the BSD license that have not been updated

File: examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c (hunk 5 above)
Comment (TanmayShah-xilinx) [reply examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c]: Ok sounds good.

File: examples/zephyr/dual_qemu_ivshmem/host/boards/pcie_ivshmem.dtsi (hunk 9 above)
[the remark this answers is not in the document]
Comment (TanmayShah-xilinx) [reply examples/zephyr/dual_qemu_ivshmem/host/boards/pcie_ivshmem.dtsi:13]: Ok. It's good to mention in README that vendor-id is random and taken from zephyr os (may be with its version number).

**Review by TanmayShah-xilinx — APPROVED** [review]
LGTM. Just can we mention vendor-id was taken from zephyr-(version) in README ? That will be good for user to understand how vendor-id is defined.

File: examples/zephyr/dual_qemu_ivshmem/host/boards/pcie_ivshmem.dtsi (hunk 9 above)
[the remark this answers is not in the document]
Comment (uLipe) [reply examples/zephyr/dual_qemu_ivshmem/host/boards/pcie_ivshmem.dtsi:13]: Put a note on README regarding the Vendor-ID

**Comment by arnopo** [comment]
@uLipe I run zephyr checkpatch.pl on your PR. there are some errors reported... could you fix them, please?

@TanmayShah-xilinx: any new concerns on this PR?

**Comment by uLipe** [comment]
@arnopo , yes sure, I will fix and run checkpatch.pl before pushing again

**Comment by uLipe** [comment]
@arnopo , @TanmayShah-xilinx  just adressed the zephyr style issues.

**Review by arnopo — COMMENTED** [review]
I still have issues with checkpatch.
could you squash the fixes and re-run  the checkpatch?

File: examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.h [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+ * @return other side destination address, -1 if none is found;
+ */
+int get_rpmsg_ivshmem_ept_dest_addr(void);
+
+/**
+ * @brief Get RPMsg-IVSHMEM, initialized, backend device for RPMSg endpoint creation
+ *
+ * @return pointer to the initialized RPMsg device
+ */
+struct rpmsg_device* get_rpmsg_ivshmem_device(void);
```
[the remark this answers is not in the document]
Comment (uLipe) [reply examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.h:25]: My bad, I skipped this file when running it.

[the remark this answers is not in the document]
Comment (uLipe) [reply examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.h]: and done!

**Review by arnopo — CHANGES_REQUESTED** [review]
Please keep a separate patch for the update of the zephyr version.  FYI in OpenAMp we use to fix the issue in the commit directly an not add fixup! on top of the series during review step

Please , could you fix the isfollowing remaining issue report by checkpatch 

CHECK: Blank lines aren't necessary after an open brace '{'
#462: FILE: examples/zephyr/dual_qemu_ivshmem/host/src/main.c:21:
+{
+

WARNING: line length of 102 exceeds 10 columns
#496: FILE: examples/zephyr/dual_qemu_ivshmem/host/src/main.c:55:
+	printf("Hello %s - Host Side, the communication over RPMsg is ready to use!\n", CONFIG_BOARD);

CHECK: Alignment should match open parenthesis
#500: FILE: examples/zephyr/dual_qemu_ivshmem/host/src/main.c:59:
+static int cmd_rpmsg_ivshmem_send(const struct shell *sh,
+			   size_t argc, char **argv)

WARNING: line length of 109 exceeds 80 columns
#516: FILE: examples/zephyr/dual_qemu_ivshmem/host/src/main.c:75:
+			printf("failed to send message over RPMsg-IVSHMEM to the Remote side: %d\n", status);


CHECK: Alignment should match open parenthesis
#534: FILE: examples/zephyr/dual_qemu_ivshmem/host/src/main.c:93:
+SHELL_STATIC_SUBCMD_SET_CREATE(sub_rpmsg_ivshmem,
+						SHELL_CMD_ARG(send, NULL,

CHECK: Alignment should match open parenthesis
#535: FILE: examples/zephyr/dual_qemu_ivshmem/host/src/main.c:94:
+						SHELL_CMD_ARG(send, NULL,
+								"Usage: rpmsg_ivshmem send <string> <number of messages>",


CHECK: Alignment should match open parenthesis
#541: FILE: examples/zephyr/dual_qemu_ivshmem/host/src/main.c:100:
+SHELL_CMD_ARG_REGISTER(rpmsg_ivshmem, &sub_rpmsg_ivshmem,
+				"Commands to use RPMsg over IVSHMEM",

WARNING: 'supress' may be misspelled - perhaps 'suppress'?
#643: FILE: examples/zephyr/dual_qemu_ivshmem/remote/prj.conf:16:
+# Enable IVSHMEM shell when debugging the driver, default OFF to supress warnings when not used.

WARNING: line length of 104 exceeds 100 columns
#705: FILE: examples/zephyr/dual_qemu_ivshmem/remote/src/main.c:53:
+	printf("Hello %s - Remote Side, the communication over RPMsg is ready to use!\n", CONFIG_BOARD);

CHECK: No space is necessary after a cast
#760: FILE: examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c:48:
+			.virt       = (void *) 0,

CHECK: Please don't use multiple blank lines
#921: FILE: examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c:209:
+
+

CHECK: Alignment should match open parenthesis
#927: FILE: examples/zephyr/dual_qemu_ivshmem/rpmsg_ivshmem_backend/rpmsg_ivshmem_backend.c:215:
+	k_thread_create(&ivshmem_ev_loop_thread,
+		ivshmem_ev_loop_stack,

**Comment by uLipe** [comment]
It is strange, my checkpatch.pl (under zephyr/scripts) did not catch these style issues, I addressed them manually btw.

**Comment by uLipe** [comment]
@arnopo , PTAL again, I ran the checkpatch using the same command you are using, the remaining issues are gone, but it stills report duplicate sign-off, I checked manually the patch and I did not find a reason for this being reported.

**Comment by uLipe** [comment]
It is strange, no matter how I align these, or change my editor settings, these three error never disappear, I even copied the alignment from other Zephyr samples and subsystems with no luck, as well trying to align mixing spaces and tabs, is there something I'm missing? @arnopo

File: examples/zephyr/dual_qemu_ivshmem/host/src/main.c [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+	return 0;
+}
+
+SHELL_STATIC_SUBCMD_SET_CREATE(sub_rpmsg_ivshmem,
+				SHELL_CMD_ARG(send, NULL,
+						"Usage: rpmsg_ivshmem send <string> <number of messages>",
+						cmd_rpmsg_ivshmem_send, 3, 0),
+				SHELL_SUBCMD_SET_END
+		);
+
```
Comment (arnopo) [inline examples/zephyr/dual_qemu_ivshmem/host/src/main.c:96]: For instance here you align only with tabs, you have also to use space to complete alignment.  
And be sure that you configure your editor using 8 spaces for a tab.
 
```c
 SHELL_STATIC_SUBCMD_SET_CREATE(sub_rpmsg_ivshmem,
-				SHELL_CMD_ARG(send, NULL,
-						"Usage: rpmsg_ivshmem send <string> <number of messages>",
-						cmd_rpmsg_ivshmem_send, 3, 0),
-				SHELL_SUBCMD_SET_END
+			       SHELL_CMD_ARG(send, NULL,
+					     "Usage: rpmsg_ivshmem send <string> <number of messages>",
+					     cmd_rpmsg_ivshmem_send, 3, 0),
+			       SHELL_SUBCMD_SET_END
 		);

 SHELL_CMD_ARG_REGISTER(rpmsg_ivshmem, &sub_rpmsg_ivshmem,
-				"Commands to use RPMsg over IVSHMEM",
-				cmd_rpmsg_ivshmem_send, 3, 0);
+		       "Commands to use RPMsg over IVSHMEM",
+		       cmd_rpmsg_ivshmem_send, 3, 0);
```
If you don't succeed, I'll correct the problem during the merge.

**Comment by uLipe** [comment]
@arnopo the problem was in my editor local config, I was able to fix the alignment warning, thank you.

**Review by arnopo — APPROVED** [review]
LGTM

**Review by TanmayShah-xilinx — APPROVED** [review]
LGTM

**Comment by TanmayShah-xilinx** [comment]
@wmamills if all look good, can we merge this change? Me and @arnopo have approved.

**Pull request merged**